trail/trail.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2016 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package trail integrates trace errors with GRPC
+//
+// Example server that sends the GRPC error and attaches metadata:
+//
+//
+//  func (s *server) Echo(ctx context.Context, message *gw.StringMessage) (*gw.StringMessage, error) {
+//     trace.SetDebug(true) // to tell trace to start attaching metadata
+//     // Send sends metadata via grpc header and converts error to GRPC compatible one
+//     return nil, trail.Send(ctx, trace.AccessDenied("missing authorization"))
+//  }
+//
+// Example client reading error and trace debug infor
+//
+//  var header metadata.MD
+//	r, err := c.Echo(context.Background(), &gw.StringMessage{Value: message}, grpc.Header(&header))
+//	if err != nil {
+//      // FromGRPC reads error, converts it back to trace error and attaches debug metadata
+//      // like stack trace of the error origin back to the error
+//		err = trail.FromGRPC(err, header)
+///     // this line will log original trace of the error
+//		log.Errorf("error saying echo: %v", trace.DebugReport(err))
+//		return
+//	}
+package trail
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+
+	"github.com/gravitational/trace"
+
+	log "github.com/Sirupsen/logrus"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+)
+
+// Send is a high level function that:
+// * logs error
+// * converts error to GRPC error
+// * attaches debug metadata to existing metadata if possible
+// * sends the header to GRPC
+func Send(ctx context.Context, err error) error {
+	meta, ok := metadata.FromContext(ctx)
+	if !ok {
+		meta = metadata.New(nil)
+	}
+	if trace.IsDebug() {
+		SetDebugInfo(err, meta)
+	}
+	log.Error(trace.DebugReport(err))
+	if len(meta) != 0 {
+		log.Infof("meta: %v", meta)
+		err2 := grpc.SendHeader(ctx, meta)
+		if err2 != nil {
+			log.Errorf("failed to send metadata: %v", err2)
```

Comment:
Sign of the method doing too much? Isn't the failure to send metadata fatal to a GRPC call?